### ISSUE TITLE
GenGEigsSolver

### ISSUE THREAD

**Issue by ajos6183** [yixuan/spectra#67]
Firstly, thank you so much for this great library.
Is there a plan to add a generalized eigen solver for general real matrices?

**Comment by ajos6183** [comment]
I actually managed to get around this requirement. So i will close this question.

**Comment by ajos6183** [comment]
Apologies for reopening this feature request. I realized I do need it.
Please let me know if there are plans to support it.

**Comment by ajos6183** [comment]
@yixuan or anyone who can help,

Is there a way of solving the sparse buckling problem (Ke.x = λ.Kg.x) where Ke is symmetric and Kg non-symmetric with this library as it stands? I tried looking through the library but could not find such a thing. I do not wish to invert any of the sparse matrices because they will end up being dense.

Sorry to ask another question but i thought i'd rephrase my question in the case that what I asked for already exists...

**Comment by yixuan** [comment]
Hi @ajos6183, I have to say sorry that right now Spectra has very limited support for generalized eigen problems, and the buckling mode has not yet been implemented.

**Comment by yixuan** [comment]
By the way, if your Ke is positive definite and Kg is symmetric, then I think I can implement such a solver shortly.

**Comment by ajos6183** [comment]
Thanks @yixuan,

Actually i noticed i had some issues with my code. It is very likely that Ke is positive definite and Kg is symmetric. This is what i have used in the past for this type of problem. I would really appreciate it if you have a better way or if you can point out any issues with my code.

```c++
int SPECTRA_SymGEigsSolver(int nev, int n, int Ke_nentries, int* Ke_RowIndices, int* Ke_ColumnPointers, double* Ke_Values,
	int Kg_nentries, int* Kg_RowIndices, int* Kg_ColumnPointers, double* Kg_Values, double* eigenvalues)
{
	Eigen::SparseMatrix<double> Ke(n, n);
	Eigen::Index Ke_colnum = 0;
	for (int i = 0; i < Ke_nentries; i++) {
		while (i >= Ke_ColumnPointers[Ke_colnum]) { Ke_colnum += 1; }
		Ke.insert(Ke_RowIndices[i], Ke_colnum - 1) = Ke_Values[i];
	}
	Eigen::SparseMatrix<double> Kg(n, n);
	Eigen::Index Kg_colnum = 0;
	for (int i = 0; i < Kg_nentries; i++) {
		while (i >= Kg_ColumnPointers[Kg_colnum]) { Kg_colnum += 1; }
		Kg.insert(Kg_RowIndices[i], Kg_colnum - 1) = Kg_Values[i];
	}

	// Construct matrix operation object using the wrapper class SparseSymMatProd
	Spectra::SparseSymMatProd<double> op(Kg);
	Spectra::SparseCholesky<double>  Bop(Ke);

	// Construct eigen solver object, requesting the largest nev eigenvalues
	Spectra::SymGEigsSolver< double, Spectra::LARGEST_MAGN, Spectra::SparseSymMatProd<double>, Spectra::SparseCholesky<double>, 0 > geigs(&op, &Bop, nev, std::min(2 * nev + 1, n));

	// Initialize and compute
	geigs.init();
	int nconv = geigs.compute();

	// Retrieve results
	Eigen::VectorXd evalues;
	if (geigs.info() == Spectra::SUCCESSFUL) { evalues = geigs.eigenvalues(); }
	for (int i = 0; i < nconv; i++) { eigenvalues[i] = evalues[i]; }
	return nconv;
}

**Comment by yixuan** [comment]
I have already implemented the buckling mode in the  [1.y.z](https://github.com/yixuan/spectra/tree/1.y.z) branch. Below is some example code:
```cpp
#include <Eigen/Core>
#include <Eigen/SparseCore>
#include <Spectra/SymGEigsShiftSolver.h>
#include <Spectra/MatOp/SymShiftInvert.h>
#include <Spectra/MatOp/SparseSymMatProd.h>
#include <iostream>

using namespace Spectra;

int main()
{
    // We are going to solve the generalized eigenvalue problem
    //     K * x = lambda * KG * x,
    // where K is positive definite, and KG is symmetric
    const int n = 100;

    // Define the K matrix, a tridiagonal matrix with 2 on the diagonal
    // and 1 on the subdiagonals
    Eigen::SparseMatrix<double> K(n, n);
    K.reserve(Eigen::VectorXi::Constant(n, 3));
    for (int i = 0; i < n; i++)
    {
        K.insert(i, i) = 2.0;
        if (i > 0)
            K.insert(i - 1, i) = 1.0;
        if (i < n - 1)
            K.insert(i + 1, i) = 1.0;
    }

    // Define the KG matrix
    Eigen::MatrixXd M = Eigen::MatrixXd::Random(n, n);
    Eigen::MatrixXd KG = M + M.transpose();

    // Construct matrix operation objects using the wrapper classes
    // K is sparse, KG is dense
    using OpType = SymShiftInvert<double, Eigen::Sparse, Eigen::Dense>;
    using BOpType = SparseSymMatProd<double>;
    OpType op(K, KG);
    BOpType Bop(K);

    // Construct generalized eigen solver object, seeking three generalized
    // eigenvalues that are closest to and larger than 1.0. This is equivalent to
    // specifying a shift sigma = 1.0 combined with the SortRule::LargestAlge
    // selection rule
    SymGEigsShiftSolver<double, OpType, BOpType, GEigsMode::ShiftInvert>
        geigs(op, Bop, 3, 6, 1.0);

    // Initialize and compute
    geigs.init();
    int nconv = geigs.compute(SortRule::LargestAlge);

    // Retrieve results
    Eigen::VectorXd evalues;
    Eigen::MatrixXd evecs;
    if (geigs.info() == CompInfo::Successful)
    {
        evalues = geigs.eigenvalues();
        evecs = geigs.eigenvectors();
    }

    std::cout << "Number of converged generalized eigenvalues: " << nconv << std::endl;
    std::cout << "Generalized eigenvalues found:\n" << evalues << std::endl;
    std::cout << "Generalized eigenvectors found:\n" << evecs.topRows(10) << std::endl;
    return 0;
}
```

Note that this new branch contains several API-breaking changes that will be included in the next major release, so some (minor) code migration is needed. It also requires C++11 now.

**Comment by ajos6183** [comment]
Thank you, I very much appreciate your assistance. 

One last question; it seems like using `GEigsMode::Cholesky` without a shift is about twice as fast as the above code for my buckling problems. Are there any advantages to using `GEigsMode::ShiftInvert` or were you just giving an example? And is there anything faster than  `GEigsMode::Cholesky` for buckling problems that is currently implemented?

**Comment by yixuan** [comment]
Your code actually solves `Kg * x = lambda * Ke * x`, not  `Ke * x = lambda * Kg * x`. The buckling mode solves the correct problem.

There was a typo in my previous code. Please replace the following lines
```cpp
SymGEigsShiftSolver<double, OpType, BOpType, GEigsMode::ShiftInvert>
    geigs(op, Bop, 3, 6, 1.0);
```
with
```cpp
SymGEigsShiftSolver<double, OpType, BOpType, GEigsMode::Buckling>
    geigs(op, Bop, 3, 6, 1.0);
```
I was trying to illustrate `GEigsMode::Buckling`, not `GEigsMode::ShiftInvert`.

**Comment by yixuan** [comment]
In summary, Cholesky mode solves `A * x = lambda * B * x`, where `B` is positive definite.
Buckling mode solves `K * x = lambda * Kg * x`, where `K` is positive definite.

**Comment by ajos6183** [comment]
Yes you are correct, i solve `Kg * x = lambda * Ke * x` and end up inverting the largest positive eigenvalues of this problem to arrive at a set of the smallest positive eigenvalues (commonly of interest in buckling problems). Without knowing the specifics, my theory is that solving `Kg * x = lambda * Ke * x` is easier than `Ke * x = lambda * Kg * x` since the positive definite matrix (Ke) is the one multiplied by the eigenvalue.

**Comment by yixuan** [comment]
I see. If you look for eigenvalues around zero then your first method that switches the role of `Ke` and `Kg` might be a better solution, since `sigma` cannot be 0 in the buckling mode.

The speed depends on many factors, for example the cost per iteration, and the convergence speed. You can use functions `num_iterations()` and `num_operations()` to see how many iterations each algorithm actually uses.

**Comment by ajos6183** [comment]
Thank you
